openverse_catalog/dags/providers/provider_api_scripts/europeana.py

```diff
@@ -8,266 +8,172 @@
-from requests.exceptions import JSONDecodeError
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
 
 
-logging.basicConfig(
-    format="%(asctime)s - %(name)s - %(levelname)s:  %(message)s", level=logging.INFO
-)
 logger = logging.getLogger(__name__)
+logging.getLogger(common.urls.__name__).setLevel(logging.WARNING)
+
+
+class EuropeanaRecordBuilder:
+    """
+    A small class to contain the record building functionality
+    and simplify testing a bit.
+    """
+
+    def get_record_data(self, data: dict) -> dict:
+        record = {
+            "foreign_landing_url": self.get_foreign_landing_url(data),
+            "image_url": data.get("edmIsShownBy")[0],
+            "foreign_identifier": data.get("id"),
+            "meta_data": self.get_meta_data_dict(data),
+            "title": data.get("title")[0],
+            "license_info": get_license_info(
+                license_url=self.get_license_url(data.get("rights"))
+            ),
+        }
+
+        data_providers = set(record["meta_data"]["dataProvider"])
+        eligible_sub_providers = {
+            s
+            for s in EuropeanaDataIngester.sub_providers
+            if EuropeanaDataIngester.sub_providers[s] in data_providers
+        }
+        if len(eligible_sub_providers) > 1:
+            raise Exception(
+                f"More than one sub-provider identified for the "
+                f"image with foreign ID {record['foreign_identifier']}"
+            )
+
+        return record | {
+            "source": (
+                eligible_sub_providers.pop()
+                if len(eligible_sub_providers) == 1
+                else EuropeanaDataIngester.providers["image"]
+            )
+        }
+
+    def get_license_url(self, license_field) -> str | None:
+        if len(license_field) > 1:
+            logger.warning("More than one license field found")
+        for license_ in license_field:
+            if "creativecommons" in license_:
+                return license_
+        return None
+
+    def get_foreign_landing_url(self, data: dict) -> str:
+        original_url = data.get("edmIsShownAt")
+        if original_url is not None:
+            return original_url[0]
+        europeana_url = data.get("guid")
+        return europeana_url
+
+    def get_meta_data_dict(self, data: dict) -> dict:
+        meta_data = {
+            "country": data.get("country"),
+            "dataProvider": data.get("dataProvider"),
+            "description": self.get_description(data),
+        }
+
+        return {k: v for k, v in meta_data.items() if v is not None}
+
+    def get_description(self, data: dict) -> str | None:
+        lang_aware_description = data.get("dcDescriptionLangAware")
+        if lang_aware_description:
+            description = lang_aware_description.get(
+                "en"
+            ) or lang_aware_description.get("def")
+        else:
+            description = data.get("dcDescription")
```

Comment:
This misses the case where `dcDescriptionLangAware` is available but does not have the `en` or `def` keys. In that case we still want to fall back to `dcDescription`.

Reply:
Ah, good catch!

Reply:
Fixed in https://github.com/WordPress/openverse-catalog/pull/821/commits/dfac506409087cc3a48d689ef49e7503bc65940b